clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java

```diff
@@ -0,0 +1,461 @@
+
+        if (spec.scope instanceof FulfillmentScope) {
+            int brokerId = ((FulfillmentScope) spec.scope).destinationBrokerId;
+            AdminApiHandler.ApiResult<K, V> result = handler.handleResponse(
+                brokerId,
+                spec.keys,
+                response
+            );
+            result.completedKeys.forEach(this::complete);
+            result.failedKeys.forEach(this::completeExceptionally);
+            result.unmappedKeys.forEach(this::unmap);
+        } else {
+            AdminApiLookupStrategy.LookupResult<K> result = keys.lookupStrategy.handleResponse(
+                spec.keys,
+                response
+            );
+            result.failedKeys.forEach(this::completeExceptionally);
+            result.mappedKeys.forEach(this::map);
+        }
+    }
+
+    /**
+     * Callback that is invoked when a `Call` is failed.
+     */
+    public void onFailure(
+        long currentTimeMs,
+        RequestSpec<K> spec,
+        Throwable t
+    ) {
+        clearInflightRequest(currentTimeMs, spec);
+        if (t instanceof DisconnectException) {
+            log.debug("Node disconnected before response could be received for request {}. " +
+                "Will attempt retry", spec.request);
+
+            // After a disconnect, we want the driver to attempt to lookup the key
+            // again (if the key is dynamically mapped). This gives us a chance to
+            // find a new coordinator or partition leader for example.
+            spec.keys.stream()
+                .filter(keys.dynamicKeys::contains)
+                .forEach(this::unmap);
+        } else {
+            spec.keys.forEach(key -> completeExceptionally(key, t));
+        }
+    }
+
+    private void clearInflightRequest(long currentTimeMs, RequestSpec<K> spec) {
+        RequestState requestState = requestStates.get(spec.scope);
+        if (requestState != null) {
+            requestState.clearInflight(currentTimeMs);
+        }
+    }
+
+    private <T extends ApiRequestScope> void collectRequests(
+        List<RequestSpec<K>> requests,
+        BiMultimap<T, K> multimap,
+        BiFunction<Set<K>, T, AbstractRequest.Builder<?>> buildRequest
+    ) {
+        for (Map.Entry<T, Set<K>> entry : multimap.entrySet()) {
+            T scope = entry.getKey();
+
+            Set<K> keys = entry.getValue();
+            if (keys.isEmpty()) {
+                continue;
+            }
+
+            RequestState requestState = requestStates.computeIfAbsent(scope, c -> new RequestState());
+            if (requestState.hasInflight()) {
+                continue;
+            }
+
+            AbstractRequest.Builder<?> request = buildRequest.apply(keys, scope);
+            RequestSpec<K> spec = new RequestSpec<>(
+                handler.apiName() + "(api=" + request.apiKey() + ")",
+                scope,
+                new HashSet<>(keys), // copy to avoid exposing mutable state
```

Comment:
line#285 also exposes `keys`. How about building an immutable copy and then pass it to both `RequestSpec` and `buildRequest`?